File: aoc22_py/day_17.py
"""The solution to day 17."""
from itertools import cycle

from .aoc_helper import Solution


class Block:
    """One of the stone shapes."""

    def __init__(self, raw: str):
        """Parse and store the raw data."""
        # Offsets of squares from the bottom left corner, y is 0 at the bottom.
        self.offsets = [
            (x, y)
            for y, line in enumerate(reversed(raw.splitlines()))
            for x, char in enumerate(line)
            if char == "#"
        ]
        self.height = 1 + max(y for _, y in self.offsets)
        self.width = 1 + max(x for x, _ in self.offsets)
        min_y_by_x: dict[int, int] = {}
        min_x_by_y: dict[int, int] = {}
        max_x_by_y: dict[int, int] = {}
        for x, y in self.offsets:
            min_y_by_x[x] = min(min_y_by_x.get(x, 99), y)
            min_x_by_y[y] = min(min_x_by_y.get(y, 99), x)
            max_x_by_y[y] = max(max_x_by_y.get(y, -1), x)
        self.lower_edge = [(x, y) for x, y in min_y_by_x.items()]
        self.left_edge = [(x, y) for y, x in min_x_by_y.items()]
        self.right_edge = [(x, y) for y, x in max_x_by_y.items()]


class Trench:
    """The trench the blocks fall into."""

    def __init__(self, width: int, jets: list[int]):
        """Set up the trench with the given width."""
        self.width = width
        self.jets = list(jets)
        self.cells: set[tuple[int, int]] = set()
        self.height: int = 0
        self.buried_height = 0
        self.jet_index = 0

    def next_jet(self) -> int:
        """Get the next jet."""
        jet = self.jets[self.jet_index]
        self.jet_index = (self.jet_index + 1) % len(self.jets)
        return jet

    @property
    def full_height(self) -> int:
        """The height of the trench, including buried cells."""
        return self.height + self.buried_height

    def advance(self, cells: set[tuple[int, int]], height: int, buried_height_increase: int):
        """Advance the simulation to the state with the given cells and height increase."""
        self.cells = cells
        self.height = height
        self.buried_height += buried_height_increase

    def fall(self, block: Block):
        """Fall the block into the trench."""
        x, y = 2, self.height
        max_x = self.width - block.width
        for _ in range(4):
            x += self.next_jet()
            x = 0 if x < 0 else max_x if x > max_x else x
        while y > 0 and self.is_clear(x, y - 1, block.lower_edge):
            y -= 1
            jet = self.next_jet()
            if self.is_clear(x + jet, y, block.left_edge if jet < 0 else block.right_edge):
                x += jet
            x = 0 if x < 0 else max_x if x > max_x else x
        for x_off, y_off in block.offsets:
            self.cells.add((x + x_off, y + y_off))
        block_height = block.height + y
        self.height = self.height if self.height > block_height else block_height

    def is_clear(self, x: int, y: int, offsets: list[tuple[int, int]]) -> bool:
        """Check if the given offsets are empty at the given coordinates."""
        for x_off, y_off in offsets:
            if (x + x_off, y + y_off) in self.cells:
                return False
        return True

    def clear_buried(self):
        """Get rid of buried cells to reduce memory usage and keep numbers small."""
        """
        cells: set[tuple[int, int]] = {(x, y) for y in range(self.height - 32) for x in range(self.width)}
        for block in BLOCKS:
            visited: set[tuple[int, int]] = set()
            open: list[tuple[int, int]] = [(x, self.height) for x in range(self.width - block.width + 1)]
            while open:
                x, y = open.pop()
                if x < 0 or x > self.width - block.width or y < 0 or y > self.height:
                    continue
                if (x, y) in visited:
                    continue
                visited.add((x, y))
                valid = True
                for x_off, y_off in block.offsets:
                    if (x + x_off, y + y_off) in self.cells:
                        valid = False
                if valid:
                    for x_off, y_off in block.offsets:
                        if (x + x_off, y + y_off) in cells:
                            cells.remove((x + x_off, y + y_off))
                    open.append((x - 1, y))
                    open.append((x + 1, y))
                    open.append((x, y - 1))
                    open.append((x, y + 1))
        if not cells:
            return
        height_buried = min(y for _, y in cells)
        self.cells = {(x, y - height_buried) for x, y in cells}
        self.buried_height += height_buried
        self.height -= height_buried
        """
        visited: set[tuple[int, int]] = set()
        open: list[tuple[int, int]] = [(x, self.height) for x in range(self.width)]
        cells: set[tuple[int, int]] = set()
        while open:
            x, y = open.pop()
            if x < 0 or x >= self.width or y < 0 or y > self.height:
                continue
            if (x, y) in visited:
                continue
            visited.add((x, y))
            if (x, y) in self.cells:
                cells.add((x, y))
            else:
                open.append((x - 1, y))
                open.append((x + 1, y))
                open.append((x, y - 1))
                open.append((x, y + 1))
        if not cells:
            return
        height_buried = min(y for _, y in cells)
        self.cells = {(x, y - height_buried) for x, y in cells}
        self.buried_height += height_buried
        self.height -= height_buried

    def __str__(self) -> str:
        """Display the trench."""
        lines: list[str] = []
        for y in range(self.height):
            line = ""
            for x in range(self.width):
                if (x, y) in self.cells:
                    line += "#"
                else:
                    line += "."
            lines.append(line)
        return "\n".join(reversed(lines))

    def copy(self) -> "Trench":
        """Return a copy of the trench."""
        trench = Trench(self.width, self.jets)
        trench.height = self.height
        trench.buried_height = self.buried_height
        trench.jet_index = self.jet_index
        trench.cells = self.cells.copy()
        return trench

    def print_with_block(self, x: int, y: int, block: Block):
        """Print the trench with the block at the given coordinates."""
        trench = self.copy()
        for x_off, y_off in block.offsets:
            trench.cells.add((x + x_off, y + y_off))
        print(trench)


RAW_BLOCKS = """
####

.#.
###
.#.

..#
..#
###

#
#
#
#

##
##
"""


BLOCKS = [Block(raw) for raw in RAW_BLOCKS.strip("\n").split("\n\n")]


class Day(Solution):
    """The solution to day 17."""

    day = 17

    def __init__(self, raw: str):
        """Parse and store the raw data."""
        self.deltas: list[int] = []
        for c in raw:
            if c == ">":
                self.deltas.append(1)
            elif c == "<":
                self.deltas.append(-1)

    def part_1(self) -> str | int | None:
        """Calculate the answer for part 1."""
        trench = Trench(7, list(self.deltas))
        for block, _ in zip(cycle(BLOCKS), range(2022)):
            trench.fall(block)
        return trench.height

    def part_2(self) -> str | int | None:
        """Calculate the answer for part 2."""
        trench = Trench(7, list(self.deltas))
        Cells = tuple[tuple[int, int], ...]
        visited: dict[tuple[int, Cells], tuple[int, int]] = {}
        remaining = 1_000_000_000_000
        while remaining > 0:
            for block in BLOCKS:
                trench.fall(block)
            remaining -= len(BLOCKS)
            if trench.jet_index == 0:
                trench.clear_buried()
                key = (trench.jet_index, tuple(sorted(trench.cells)))
                if key in visited:
                    remaining_then, height_then = visited[key]
                    period = remaining_then - remaining
                    increment = trench.full_height - height_then
                    periods, remaining = divmod(remaining, period)
                    trench.buried_height += periods * increment
                    break
                visited[key] = remaining, trench.full_height
        for block, _ in zip(cycle(BLOCKS), range(remaining)):
            trench.fall(block)
        return trench.full_height


if __name__ == "__main__":
    Day.submit()

File: aoc22_rs/src/days/day17.rs
use rustc_hash::{FxHashMap, FxHashSet};

use crate::{vec2::Vec2, Day};

#[derive(Clone)]
pub struct Day17(Vec<Jet>);

#[derive(Clone, Copy)]
enum Jet {
    Left,
    Right,
}

impl Jet {
    const fn as_delta(self) -> Vec2 {
        match self {
            Self::Left => Vec2::new(-1, 0),
            Self::Right => Vec2::new(1, 0),
        }
    }
}

#[derive(Debug)]
struct Block {
    offsets: Vec<Vec2>,
    height: isize,
    width: isize,
    lower_edge: Vec<Vec2>,
    left_edge: Vec<Vec2>,
    right_edge: Vec<Vec2>,
}

impl Block {
    fn new(offsets: Vec<Vec2>) -> Self {
        let height = offsets.iter().map(|v| v.1).max().unwrap() + 1;
        let width = offsets.iter().map(|v| v.0).max().unwrap() + 1;
        let lower_edge = (0..width)
            .map(|x| {
                offsets
                    .iter()
                    .filter(|v| v.0 == x)
                    .min_by_key(|v| v.1)
                    .unwrap()
            })
            .copied()
            .collect();
        let left_edge = (0..height)
            .map(|y| {
                offsets
                    .iter()
                    .filter(|v| v.1 == y)
                    .min_by_key(|v| v.0)
                    .unwrap()
            })
            .copied()
            .collect();
        let right_edge = (0..height)
            .map(|y| {
                offsets
                    .iter()
                    .filter(|v| v.1 == y)
                    .max_by_key(|v| v.0)
                    .unwrap()
            })
            .copied()
            .collect();
        Self {
            offsets,
            height,
            width,
            lower_edge,
            left_edge,
            right_edge,
        }
    }

    fn from_raw(raw: &str) -> Self {
        let offsets = raw
            .lines()
            .rev()
            .enumerate()
            .flat_map(|(y, line)| line.chars().enumerate().map(move |(x, c)| (x, y, c)))
            .filter(|(_, _, c)| *c == '#')
            .map(|(x, y, _)| Vec2::new(x as isize, y as isize))
            .collect();
        Self::new(offsets)
    }
}

const RAW_BLOCKS: &str = "####

.#.
###
.#.

..#
..#
###

#
#
#
#

##
##";

lazy_static::lazy_static! {
    static ref BLOCKS: Vec<Block> = RAW_BLOCKS.split("\n\n").map(Block::from_raw).collect();
}

const WIDTH: isize = 7;
const DOWN: Vec2 = Vec2::new(0, -1);

struct Trench<'a> {
    jets: &'a [Jet],
    cells: FxHashSet<Vec2>,
    height: isize,
    buried_height: isize,
    jet_index: usize,
}

impl<'a> Trench<'a> {
    fn new(jets: &'a [Jet]) -> Self {
        Trench {
            jets,
            cells: FxHashSet::default(),
            height: 0,
            buried_height: 0,
            jet_index: 0,
        }
    }

    #[inline]
    fn next_jet(&mut self) -> Jet {
        let jet = self.jets.get(self.jet_index).copied().unwrap();
        self.jet_index += 1;
        if self.jet_index >= self.jets.len() {
            self.jet_index = 0;
        }
        jet
    }

    const fn full_height(&self) -> isize {
        self.height + self.buried_height
    }

    fn fall(&mut self, block: &Block) {
        let mut position = Vec2::new(2, self.height);
        let max_x = WIDTH - block.width;
        for _ in 0..4 {
            position += self.next_jet().as_delta();
            position.0 = position.0.clamp(0, max_x);
        }
        while position.1 > 0 && self.is_clear(position + DOWN, &block.lower_edge) {
            position += DOWN;
            let jet = self.next_jet();
            let edge = match jet {
                Jet::Left => &block.left_edge,
                Jet::Right => &block.right_edge,
            };
            let new_position = position + jet.as_delta();
            if self.is_clear(new_position, edge) {
                position = new_position;
                position.0 = position.0.clamp(0, max_x);
            }
        }
        for &offset in &block.offsets {
            self.cells.insert(position + offset);
        }
        self.height = self.height.max(block.height + position.1);
    }

    fn is_clear(&self, position: Vec2, offsets: &[Vec2]) -> bool {
        offsets
            .iter()
            .all(|&offset| !self.cells.contains(&(position + offset)))
    }

    fn clear_buried(&mut self) {
        let mut visited = FxHashSet::default();
        let mut open: Vec<Vec2> = Vec::new();
        for x in 0..WIDTH {
            open.push(Vec2::new(x, self.height - 1));
        }
        let mut cells = FxHashSet::default();
        while let Some(next) = open.pop() {
            if next.0 < 0 || next.0 >= WIDTH || next.1 < 0 || next.1 >= self.height {
                continue;
            }
            if !visited.insert(next) {
                continue;
            }
            if self.cells.contains(&next) {
                cells.insert(next);
            } else {
                open.push(next + Vec2::new(-1, 0));
                open.push(next + Vec2::new(1, 0));
                open.push(next + Vec2::new(0, -1));
            }
        }
        if cells.is_empty() {
            return;
        }
        let height_buried = cells.iter().map(|v| v.1).min().unwrap();
        self.cells = cells
            .iter()
            .map(|&v| v - Vec2::new(0, height_buried))
            .collect();
        self.buried_height += height_buried;
        self.height -= height_buried;
    }
}

impl Day for Day17 {
    fn parse(input: &str) -> Self {
        Self(
            input
                .chars()
                .map(|c| match c {
                    '<' => Jet::Left,
                    '>' => Jet::Right,
                    _ => unreachable!(),
                })
                .collect(),
        )
    }

    fn part1(&self) -> String {
        let mut trench = Trench::new(&self.0);
        for block in BLOCKS.iter().cycle().take(2022) {
            trench.fall(block);
        }
        trench.full_height().to_string()
    }

    fn part2(&self) -> String {
        let mut trench = Trench::new(&self.0);
        let mut visited = FxHashMap::default();
        let mut remaining = 1_000_000_000_000;
        while remaining >= BLOCKS.len() as isize {
            for block in BLOCKS.iter() {
                trench.fall(block);
            }
            remaining -= BLOCKS.len() as isize;
            //    if trench.jet_index == 0 {
            trench.clear_buried();
            let mut key: Vec<_> = trench.cells.iter().copied().collect();
            key.sort();
            if let Some((remaining_then, height_then)) =
                visited.insert((trench.jet_index, key), (remaining, trench.full_height()))
            {
                let period = remaining_then - remaining;
                let increment = trench.full_height() - height_then;
                let cycles = remaining / period;
                remaining %= period;
                trench.buried_height += cycles * increment;
                break;
                //               }
            }
        }
        for block in BLOCKS.iter().cycle().take(remaining as usize) {
            trench.fall(block);
        }
        trench.full_height().to_string()
    }
}
